[Discover] Supports SQL query language (#134429)

stratoula opened the pull request:
## Summary

This PR enables the SQL queries in Discover. The feature is available after an advanced setting is enabled.
<img width="943" alt="image" src="https://user-images.githubusercontent.com/17003240/179972522-e5b647cd-974b-46ab-88d8-bd1fcaade15f.png">

The user can run sql queries and also save them as saved searches. These searches cant be used for creating an aggregation based visualization but can be imported on a dashboard.

![sql_discover](https://user-images.githubusercontent.com/17003240/176728316-bd02becd-3458-469e-aba9-564f44dece1c.gif)

![sql_discover1](https://user-images.githubusercontent.com/17003240/176730350-024f50b3-f7c1-4708-836a-b3fc6b75ac0c.gif)

For the technical preview we are going to support the following features:
- SQL integration, FROM clause bound to a data view (error if no match)
- No autocomplete
- No date histogram
- Saving as part of the saved search
- No switch to Lens
- No filter by cell click
- Time range integration
- Dashboard filter integration
- No async search / search sessions
- No “add field” button
- No document viewer / context viewer 
- Inline documentation

### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [x] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

stratoula commented:
I am currently working on these design improvements https://github.com/elastic/kibana/pull/134429#pullrequestreview-1043580227

elasticmachine commented:
Pinging @elastic/kibana-vis-editors @elastic/kibana-vis-editors-external (Team:VisEditors)

stratoula commented:
@MichaelMarcialis I have addressed the majority of your comments except from:
- > Is there an alternative code highlighting theme that we can use which doesn't so prominently utilize blue colored text? I'm worried the current blue color is very similar to the color for links that we use across Kibana.

You can create your custom theme with monaco so we could but this theme is used in Lens, Vega and in other places. Why should we differentiate this one? 

- > The red underlining of errors within the editor appears odd currently. For example, when supplying an unavailable data view with the SQL FROM clause, only the R in FROM will be underlined in red. It seems odd that only a single character in the clause is being underlined. Further, would it be better to have the unavailable data view underlined here instead?

I can't replicate this. Can you give me the query that you are using?

- > I had suggested in the [wireframes](https://whimsical.com/wireframes-v3-sql-5xNkDTdWTkJ28oJTce7HzX@VsSo8s35Uzyeqa1mLRRsdM) that clicking an individual error in the error list popover should focus the user's cursor on that error in the editor. That doesn't appear to be implemented here. Would it be possible to add?

I didn't implement it as we can only have one error and not a list of errors. So the user can't be confused with it. Focusing on a specific line of the monaco editor is also something that needs investigation. I can add it on the list for 8.5.

- > Can we apply the standard rounded corners to the top-left and bottom-left corners of the editor to better match other EUI inputs?

- > Would it be possible to apply some vertical padding at the beginning and end of the focused editor? I'm thinking 8px will be sufficient.

These things are super tricky with the monaco editor unfortunately. I would like to investigate them on a follow up PR


- > The single visible line of query text is not vertically centered in its container currently. Would it be possible to vertically center this text?

This is something that I had investigated when I started working with the monaco editor. The only thing we can do is to reduce the height but this creates other problems. I could investigate further but I might need help on that


- > When SQL errors are present and the query is run, the current error messaging doesn't currently appear as suggested in the [wireframes](https://whimsical.com/wireframes-v3-sql-5xNkDTdWTkJ28oJTce7HzX@7YNFXnKbYrKKFwhNjN7DT). The field list still shows (even in situations where a non-existing data view is being referenced) and the only option provided to users is to show the error message in a toast. Would it be possible to change this to come closer to the what the wireframes suggest?

I used the current Discover logic for displaying errors. I can look into it on a follow up PR!

Depending on my comments above and the fact that this feature is hidden under an advanced setting, do we consider them as blockers for merging this PR?

ThomThomson reviewed (approved):
Presentation team dashboard typing changes LGTM!

kertal reviewed (commented):
What a huge PR in every way 🎉 ! I've got a question while testing it, we intend also to show SQL on a dashboard, right? because this seems to be currently not implemented:

https://github.com/elastic/kibana/blob/92a46f53444c6e419967afa55b2e7664b7c9615a/src/plugins/discover/public/embeddable/saved_search_embeddable.tsx#L210

Which leads to the issue that document data is displayed instead of the SQL one.

stratoula commented:
Thanx @kertal for the fix!! Also I allow filters now on the dashboard level for sql saved searches https://github.com/elastic/kibana/pull/136702/commits/0a91517386a77bb7d2dfeeb6bbc1d044634d5e14

nreese reviewed (approved):
kibana-gis changes LGTM
code review

jloleysens commented on packages/kbn-es-query/src/es_query/build_es_query.ts:
```diff
@@ -44,7 +45,7 @@ function removeMatchAll<T>(filters: T[]) {
 export function buildEsQuery(
   indexPattern: DataViewBase | undefined,
-  queries: Query | Query[],
+  queries: Query | AggregateQuery | Array<Query | AggregateQuery>,
```
it may be useful to create a top-level type like:

```
type AnyQuery = Query | AggregateQuery
```

And reference that everywhere instead.

jloleysens reviewed (commented):
I've reviewed the code in `es-query` and `data` I think it'd be good if @lukasolson takes a look at the unified search section (not sure whether we want to block on that though.

Tested this out locally and it looks like I'm getting some weird input behaviour. Is this an issue in the monaco editor?


https://user-images.githubusercontent.com/8155004/180253315-fc0441d5-3edc-473e-bb8f-2b3433ea7782.mov

jloleysens commented:
Overall, changes are looking good though, great work @stratoula

stratoula commented:
@jloleysens thanx a lot for your review! I address all your comments. About the behavior on the monaco editor I cant replicate it. I changed something though on the monaco configuration, can you test if it solves the problem?

kertal commented on src/plugins/discover/public/application/main/components/top_nav/on_save_search.tsx:
```diff
@@ -124,13 +130,15 @@ export async function onSaveSearch({
       copyOnSave: newCopyOnSave,
       isTitleDuplicateConfirmed,
     };
+    const navigateOrReloadSavedSearch = !Boolean(onSaveCb);
```
just for my understanding, onSaveBb is undefined if I switch away from SQL mode with saving the saved search, but not navigating to the new saved search, right?

stratoula replied:
Correct! Is for the modal (Switch with saving first)

kertal replied:
thx 👍

kertal commented on src/plugins/discover/public/application/main/hooks/use_discover_state.ts:
```diff
@@ -226,6 +249,41 @@ export function useDiscoverState({
     }
   }, [initialFetchStatus, refetch$, indexPattern, savedSearch.id]);
 
+  const fetchResults = useCallback(() => {
```
Nit: I think this name might be misleading, maybe `getResultColumns` might be more descriptive?

stratoula commented on packages/kbn-es-query/src/filters/build_filters/types.ts:
```diff
@@ -82,6 +82,12 @@ export type Query = {
 };
 
+// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
+export type AggregateQuery = {
```
This is a nice idea but I would like to investigate it further when we add the esql. I just added it there for reference and it is what we had agreed on the RFC. (If I change it now, I will have some type failures so I would prefer to look into it later!

stratoula replied:
This is a nice idea but I would like to investigate it further when we add the esql support. I use it here as a reference as it is the type that was proposed on the RFC. If I change it now it will create many types failures so I would prefer to investigate it later!

stratoula replied:
I will look into it further

stratoula replied:
Thanx for that, this is how it was suggested on the RFC but I missed it somehow 😓 Addressed!

kertal commented on src/plugins/discover/public/application/main/hooks/use_discover_state.ts:
```diff
@@ -226,6 +249,41 @@ export function useDiscoverState({
+  useEffect(() => {
+    async function fetchDataview() {
+      if (state.query && isOfAggregateQueryType(state.query)) {
+        const indexPatternFROMQuery = getIndexPatternFromSQLQuery(state.query.sql);
+        const idsTitles = await indexPatterns.getIdsWithTitle();
+        const dataViewObj = idsTitles.find(({ title }) => title === indexPatternFROMQuery);
```
Nit, because if tend to read FROMQuery ... something like a SQLQuery 😄 
```suggestion
        const indexPatternFromQuery = getIndexPatternFromSQLQuery(state.query.sql);
        const idsTitles = await indexPatterns.getIdsWithTitle();
        const dataViewObj = idsTitles.find(({ title }) => title === indexPatternFromQuery);
```

kertal reviewed (commented):
Review part II, added some nits, and would have continued to a LGTM, but I think I found a hopefully last issue. 

* So you're editing a SQL and want to switch to a 'No-SQL' data view
* You're getting a warning: Current text-based query will be cleared
* You select `Don't show this warning again`
* You click Switch without saving
* Next time you switch to SQL, you can no longer switch back to `No-SQL`

stratoula commented on src/plugins/discover/public/application/main/components/top_nav/on_save_search.tsx:
```diff
@@ -90,13 +94,15 @@ export async function onSaveSearch({
   services: DiscoverServices;
   state: GetStateReturn;
   onClose?: () => void;
+  onSaveCb?: () => void;
```
I used the onSaveCb because later on the function you declare another onSave function

stratoula commented:
@kertal thanx for the bug! I think I had tested this scenario only once 😓 Fixed!

kertal commented:
> You selected that you dont want to see this warning again? Bad for you, you will be on sql mode forever :D
> @kertal thanx for the bug! I think I had tested this scenario only once 😓 Fixed!

Caught in the act, I feel guilty! You know I wasn't unhappy in this state 😄 , but I was thinking some users might complain about this trap 😈 ... finish the review Friday morning ... thx for the ultra fast fix

davismcphee reviewed (commented):
I got a chance to play around with this yesterday and today and I love the feature, especially all the different functions we have access to! I didn't look through the code much yet, but I tested the UI pretty thoroughly and came across some things we may want to address. I think some of them are more important fixes, but a lot are just nits or things I may not understand. Also sorry in advance for all the screenshots and videos! I figured it would be easier to explain where most are UI related.

### Editor

* There doesn't seem to be a scrollbar in the SQL editor like there is with our other Monaco editors.
* Undo is not supported more than one step, which is different from our other Monaco editors. <video src="https://user-images.githubusercontent.com/25592674/180303943-b04a9f34-ea80-4d79-901a-c92a9f3a0e9e.mov"></video>
* Putting a semicolon at the end of my SQL statement caused a failure for me. This might be expected, but some of the examples in the SQL reference include semicolons, so I thought they'd be permitted. <video src="https://user-images.githubusercontent.com/25592674/180304191-dbaf71c3-ddce-4a9d-91cc-66c9db160a66.mov"></video>
* The SQL error tooltip is obstructed in collapsed view.
![error_tooltip](https://user-images.githubusercontent.com/25592674/180304630-943c9b60-9e99-45e7-8d92-750a03cdb59f.png)
* With longer multiline queries, the final line of SQL is cut off in the editor (both expanded and collapsed modes).
![sql_cut_off](https://user-images.githubusercontent.com/25592674/180304682-17e0833b-7e5a-4043-a12d-a711d8edd538.png)
* The red line seems to ignore double quotes and other symbols. <video src="https://user-images.githubusercontent.com/25592674/180304752-3fea71c3-7df4-4ff7-8601-ef9553c75617.mov"></video>
* Highlighting and removing multiple lines in the editor doesn't update the height, but one line at a time does. <video src="https://user-images.githubusercontent.com/25592674/180304968-c0ca2fc6-2e7f-4abb-9fd4-f9c1e6bed902.mov"></video>
* When editing a query with an error, the red line partially stays from where you begin editing — it seems like this might be expected but I'm not sure. <video src="https://user-images.githubusercontent.com/25592674/180305039-00e19834-aace-4417-8d96-c22b71aa7b6a.mov"></video>

### History/Navigation

* Navigating from Discover to SQL mode pushes two extra history entries that behave strangely — after first back click, it clears the grid; after the second back click, it fixes the grid but remains in SQL mode; then after the third back click, it returns to Discover. <video src="https://user-images.githubusercontent.com/25592674/180305177-f7d6883f-453b-4325-ad4a-35dd248347c3.mov"></video>
* Navigating from SQL mode back to a data view also seems to push an extra history entry. This doesn't seem to happen when using the breadcrumb. <video src="https://user-images.githubusercontent.com/25592674/180305257-6bdd6362-9653-4642-95da-ad16222cd99c.mov"></video>
* The 'query will be cleared' modal doesn't appear for me when navigating back to Discover via the breadcrumb link or the open search flyout. <video src="https://user-images.githubusercontent.com/25592674/180305531-b665be4f-f803-4538-998b-34b644e017da.mov"></video>
* When clicking 'Save and switch' in the 'query will be cleared' modal, then clicking 'Cancel' in the 'Save search' modal, it leaves SQL mode instead of cancelling the whole action which was surprising to me. <video src="https://user-images.githubusercontent.com/25592674/180305704-e40dd01e-6d53-4024-b09c-9900ca2927df.mov"></video>

### Grid

* After querying an invalid data view, selected fields stay in sidebar — refresh the page and they disappear. <video src="https://user-images.githubusercontent.com/25592674/180305799-fee52180-54c6-46f4-ba39-3931cc13b84f.mov"></video>
* Sometimes when navigating from Discover to SQL mode, it would default to showing all columns, while other times it would default to only the 'Document' column, even though it seemed like my state was identical. I'm not sure if this is a bug or not, but I wasn't able to figure out how or why this was happening. <video src="https://user-images.githubusercontent.com/25592674/180305878-1be48f68-ead4-44e0-9fb8-548dbba9eadf.mov"></video>

### SQL Reference
* The SQL reference isn't very usable in mobile view. I don't know if this really matters, but maybe using a flyout instead could solve this issue?
![mobile_sql_reference](https://user-images.githubusercontent.com/25592674/180306033-2fed3999-f0f6-4c10-8452-aea9c2eb6f40.png)
* Looks like there's a small typo in the SQL reference.
![typo](https://user-images.githubusercontent.com/25592674/180306110-31715aa6-b3b7-4434-baf6-e724ecc49210.png)

As far as the actual SQL support, it worked great for me. I like the UX and the feature seems really useful, so I think users will like this. I'll take a look at the code tomorrow and leave my review as well.

MichaelMarcialis reviewed (commented):
Wow, you got those initial design changes in fast! Thanks, @stratoula! I gave it a second look today and have compiled my comments below. Let me know if you have any questions or require any additional design support.

# Previous Comments

## Confirmation Modal

> When choosing to "Save and switch", a second modal for saving the search appears on top of the confirmation modal. Could we replace the confirmation modal with the save modal, rather than stacking them? At that point, completing the save modal should then perform the switch. Canceling out of the save modal should be considered the user wanting to cancel and cancel the entire mode switch action.

It looks like we're no longer stacking the two modals, which is awesome. The only remaining issue is that that using the `X` or "Cancel" buttons in the save modal cancels the save but still switches the user to data view mode. Could they instead cancel the save and keep them in SQL mode? I think that would be safer and preferable for users.

## Editor

### General

> Can we reduce the width of the line number column in the current error list popover? It feels overly wide currently.

It looks like we've reduced the width down to the width of the line text itself. It's indeed more narrow, but perhaps a tad too much in this case. To ensure a comfortable (but not excessive) space between the line count and error message columns, could we set the gutter space between them to be 32px (i.e. `gutterSize="xl"` for that `EuiFlexGroup`)? And while we're in there, could we also remove the center alignment so that the text is nicely aligned to the top in both columns?

![image](https://user-images.githubusercontent.com/3884767/180310405-8241b23d-4a77-4a9c-8078-071d0afb997b.png)

> Can we add a light gray background behind the line numbers on the left of the editor for added separation between them and the query?

This addition looks great. Could we possibly tweak some of the colors to finish it off? For the background, could we try `$euiColorLightestShade`? Also, for the actual line number text colors, can we use `$euiColorDisabledText` for the line numbers by default and `$euiTextSubduedColor` for the currently focused line?

> Let's change the "⌘ + Enter" shortcut from being in an `EuiBadge` to `EuiCode` (to better match how we display keyboard shortcuts elsewhere in Kibana).

Thanks for this change. Can we also add a `transparentBackground` prop to the `EuiCode` component to keep the two different shades of gray (code background and footer background) from clashing?

### Compact/auto-sizing editor

#### Unfocused

> When more than one line of SQL is entered into the editor and the editor is then unfocused, the original design expectation was that the query would be minified down into a single line (i.e. line breaks turned into spaces), with as much of the multi-line query being shown as possible until reaching the rightmost edge of the available space. Only at that point should the query text-overflow to ellipsis. However, what appears to be happening in the PR is that only the first line with appended ellipsis is shown when the editor is unfocused, regardless of length. Can we change to the original intended behavior?

This appears to be working largely as intended now. The only remaining issue I see is that the minification via adding a space between line breaks is actually adding two spaces (`&nbsp;&nbsp;`). Can we make this a single space to avoid showing a double space in place of these line breaks when the editor is unfocused?

> The bottom border that appears below the editor to indicate either focus (blue) or invalid (red) states appears to be 1px here. Could we change these to be 2px to match the size of similar states in other EUI inputs?

It looks like the unfocused invalid state still shows as a 1px border. Possible to increase to 2px to match elsewhere?

![image](https://user-images.githubusercontent.com/3884767/180310814-96ef8a5e-2659-42a7-a054-f361cab23881.png)

#### Focused

> When the editor is focused, a scrollbar track for the editor appears to the right, even if the maximum height for the editor has yet to be exceeded. Could we change this behavior to hide the scrollbar track until the user has the need/ability to scroll the editor?

It looks like the scrollbar track still shows prior to text vertically overflowing the editor. Also, there now appears to be a bit of space between the top of the scrollbar track and the top of the editor.

![image](https://user-images.githubusercontent.com/3884767/180310987-515fe6af-de0d-45a4-8a17-2bc4c551c9a3.png)

### Expanded/manual sizing editor

> Can we add additional padding to both the header (word wrap, compact, SQL reference) and footer (line count, error count, and run query shortcut) areas? I'm thinking a total of 8px vertically and 16px horizontally should look nice for both.

It appears that the footer received some padding, but the header was missed. Would it be possible to add there as well?

> Both the the header (word wrap, compact, SQL reference) and footer (line count, error count, and run query shortcut) areas appear misaligned on the right side. Can this be fixed?

This looks better overall, but it still looks like the footer is ~1px off on the right and top. Would it be possible to fix that?

![image](https://user-images.githubusercontent.com/3884767/180311274-3dfc2bab-cf60-4b03-96f1-9672f3ea706d.png)

> The bottom border that typically appears below EUI inputs denoting focus and invalid states is missing when the editor is expanded. Can we add those in?

Thanks for adding these in. The only remaining issue I noticed is that the invalid state's red border is only showing when the editor is focused (when unfocused, it disappears). I believe we typically show the invalid red border for as long as the input is invalid, regardless if focused or not. Can we mimic that behavior here?

> It appears that a custom resizer is being used when the editor is in expanded mode. Would it be possible to switch to using EUI's resizable container components to ensure consistent styling?

You previously mentioned that we wouldn't be able to use EUI's resizer because the unified search and Discover are separate plugins. You also mentioned that you can't use the `EuiResizableButton` component because it's not exported from EUI. @elastic/eui-design, is it possible to make that component exportable for our use here? Otherwise, if we absolutely can't use EUI's resizable components, can we at least extend (or copy, in the worst case) the styles to give it the same visual treatment?

# New Comments

- The word wrap functionality appears broken currently and doesn't work in either compact or expanded modes.

- Can we increase the space between the line and error counts in the editor footer to be 16px?

- Can we increase the space between the compact and SQL reference buttons in the expanded editor header to be 8px?

- There appears to be a new border between the unified search UI and page content. If it's not part of a separate issue in unified search, can we remove consider removing it? The unified search UI is fairly dense and horizontally linear already and I feel it doesn't really need the extra border for framing. Removing it may help cut down on some visual noise there as well.

- Currently, it seems that error checking and underlining only occurs after a query is ran. Is this intentional? Or should we perform error checking after a pause in typing and/or set time intervals (similar to how the formula editor works in Lens)?

MichaelMarcialis commented:
@stratoula: Some replies to your previous comments above:

> * > The red underlining of errors within the editor appears odd currently. For example, when supplying an unavailable data view with the SQL FROM clause, only the R in FROM will be underlined in red. It seems odd that only a single character in the clause is being underlined. Further, would it be better to have the unavailable data view underlined here instead?
> 
> I can't replicate this. Can you give me the query that you are using?

Interesting, I appear to be unable to replicate it now as well. Perhaps is something that was fixed along the way, or I was just losing my mind. Feel free to ignore!

> * > I had suggested in the [wireframes](https://whimsical.com/wireframes-v3-sql-5xNkDTdWTkJ28oJTce7HzX@VsSo8s35Uzyeqa1mLRRsdM) that clicking an individual error in the error list popover should focus the user's cursor on that error in the editor. That doesn't appear to be implemented here. Would it be possible to add?
> 
> I didn't implement it as we can only have one error and not a list of errors. So the user can't be confused with it. Focusing on a specific line of the monaco editor is also something that needs investigation. I can add it on the list for 8.5.

Sounds good. Thanks!

> * > When SQL errors are present and the query is run, the current error messaging doesn't currently appear as suggested in the [wireframes](https://whimsical.com/wireframes-v3-sql-5xNkDTdWTkJ28oJTce7HzX@7YNFXnKbYrKKFwhNjN7DT). The field list still shows (even in situations where a non-existing data view is being referenced) and the only option provided to users is to show the error message in a toast. Would it be possible to change this to come closer to the what the wireframes suggest?
> 
> I used the current Discover logic for displaying errors. I can look into it on a follow up PR!

Cool. Thanks for looking into that.

> Depending on my comments above and the fact that this feature is hidden under an advanced setting, do we consider them as blockers for merging this PR?

I personally don't consider any of these three items blockers.

thompsongl commented:
> You previously mentioned that we wouldn't be able to use EUI's resizer because the unified search and Discover are separate plugins. You also mentioned that you can't use the EuiResizableButton component because it's not exported from EUI. https://github.com/orgs/elastic/teams/eui-design, is it possible to make that component exportable for our use here? Otherwise, if we absolutely can't use EUI's resizable components, can we at least extend (or copy, in the worst case) the styles to give it the same visual treatment?

Could you open an issue in the EUI repo with the use case you have? Probably just makes sense to export the component, but we can discuss over there.

gchaps commented on src/plugins/discover/server/ui_settings.ts:
```diff
@@ -303,4 +308,26 @@ export const getUiSettings: (docLinks: DocLinksServiceSetup) => Record<string, U
+    value: false,
+    description: i18n.translate('discover.advancedSettings.enableSQLDescription', {
+      defaultMessage:
+        '{technicalPreviewLabel} This tech preview feature is highly experimental--do not rely on this for production saved searches or dashboards. This setting enables SQL as a text-based query language in Discover. If you have feedback on this experience please reach out to us on {link}',
```
In other places in the UI, we use a "Technical preview" label with a tooltip.  Here is a screenshot with our standard wording.  Can we do something like that here?

![image (15)](https://user-images.githubusercontent.com/33642766/180327492-786a775b-9e89-43e5-aede-3cfc0ef9f1e7.png)

The setting label could then be :

**SQL queries in Discover**

Enable SQL as a text-based query language in Discover.  Do not use for saved searches and dashboards in a production environment. Provide feedback on our [community page](https://discuss.elastic.co/c/elastic-stack/kibana/7).

stratoula replied:
I am afraid I can't add a tooltip there Gail. It is an implementation limitation.

stratoula commented:
@MichaelMarcialis I addressed most of your comments. Some comments from my side:
- About the word wrap I used the monaco's functionality that works like Lens. It works for me except I am missing something
- > There appears to be a new border between the unified search UI and page content. If it's not part of a separate issue in unified search, can we remove consider removing it? The unified search UI is fairly dense and horizontally linear already and I feel it doesn't really need the extra border for framing. Removing it may help cut down on some visual noise there as well.

This is not the result of my PR, it was added by Caroline during the initial phase of the Unified search project

- > Currently, it seems that error checking and underlining only occurs after a query is ran. Is this intentional? Or should we perform error checking after a pause in typing and/or set time intervals (similar to how the formula editor works in Lens)?

Currently the query is executed when the user clicks the submit button. We intend to have this but not on the first phase, it is not trivial.

- About the scrollbars it is annoying I know but I cant make it work as we want easily. I have to look into it (I fixed the padding issue though)
- About the resizer I am fine on copying the styles but it is a bit confusing for me as I am not very comfortable with the EUI code. I would need some help here.

jloleysens reviewed (approved):
Great work @stratoula !

I'm approving for the `es-query` and `data/common` changes for now! It'd still be good to get eyes from @lukasolson at some point. Perhaps not necessary to block on that though.

stratoula commented:
@davismcphee thanx a ton for your review. I will comment on the things I havent fixed or things I have fixed and need a comment:
- scrollbars, I hide them but the user can still scroll. I want to change on a follow up PR the way they function as Michael has proposed but chaning the scrollbars behavior in monaco in painful. As I said it will be on a follow up PR. I dont consider it as a blocker
- Undo, also need to investigate, I cant figure out why this is broken. I see from the monaco forum that it is quite sensitive and we have hacked the monaco behavior with every possible way so I guess it has broken due to that. Need to investigate on a follow up PR. I dont consider it as a blocker.
- The tooltip! I couldn't fix it with overwritting the monaco CSS. What I did is to display it only on the expanded mode
- About the cutoff lines, this is super annoying and it is caused by me adding some paddings on the editor. It happens when the editor overpasses the max height set. There is a `top` property in monaco that is not computed well after my paddings. I can overwrite it, it is an internal monaco calulation. So I reverted my paddings cc @MichaelMarcialis. I  am so sorry, I cant find any way to add paddings without causing the bug that Davis mentions. If you could take a look and help me on that it would be really appreciated!
- > Highlighting and removing multiple lines in the editor doesn't update the height, but one line at a time does. removing_lines.mov 
I can't replicate it 🤔 

-> When editing a query with an error, the red line partially stays from where you begin editing — it seems like this might be expected but I'm not sure
We want to improve this behavior in general by informing the user while typing. I think it is fine for now

- About the navigation I discussed it with @kertal and we agreed for the discovery team to look into it on a follow up PR. 

- > After querying an invalid data view, selected fields stay in sidebar 

I mention it somewhere but I am going to deal with it on another PR! 

- > Sometimes when navigating from Discover to SQL mode, it would default to showing all columns, while other times it would default to only the 'Document' column, even though it seemed like my state was identical. I'm not sure if this is a bug or not, but I wasn't able to figure out how or why this was happening

I can't replicate this 

- > The 'query will be cleared' modal doesn't appear for me when navigating back to Discover via the breadcrumb link or the open search flyout. continue_without_saving.mov 

Yes also something I have on my list but this needs a refactoring to export the modal from the unified search plugin and call it onAppLeave on the discover app. I prefer to do it on a follow up PR!

stratoula commented:
@MichaelMarcialis @thompsongl I want to explain with more details while I think I cant use the resizable component. Maybe I am wrong and you can help with that. On the gif you can see how we want this to work
![unified_search_resizer](https://user-images.githubusercontent.com/17003240/180419687-c2b0de82-f392-4778-9164-cafd5fc7236d.gif)

![Discover-Elastic](https://user-images.githubusercontent.com/17003240/180419544-6d7d0d07-08d0-4f54-8d61-2a312d4de915.png)

- The resizable panel 1 is one of the components of the unified_search plugin and is actually the monaco editor on the extended mode (I dont want to be resized on the initial or the compact mode)

- The resizable panel 2 doesnt live on this plugin. It is actually the rest of the page.

Here is the editor implementation https://github.com/elastic/kibana/blob/unified-search-text-based-lang/src/plugins/unified_search/public/query_string_input/text_based_languages_editor/index.tsx#L484

and here is my custom implementation of the draggable button 
https://github.com/elastic/kibana/blob/unified-search-text-based-lang/src/plugins/unified_search/public/query_string_input/text_based_languages_editor/index.tsx#L591

I really cant see on how I can use the EuiResizableContainer here. If I am missing something, I would appreciate some help here.

flash1293 commented:
> I really cant see on how I can use the EuiResizableContainer here. If I am missing something, I would appreciate some help

I don't have a simple idea either how to do this - it's annoying that it's such a hard problem for how the components are structured. The only real solution I'm seeing is to change the way the unified search component is embedded into the surrounding app - instead of a component in a separate branch in the tree it's the root of the tree and the actual app content is rendered as children of the unified search bar. This way, the unified search bar can wrap the app content in a way it makes sense. However this is a pretty fundamental change for all the apps using the unified search bar. We need to give this more thought, I would like to split it out of the base feature.

kertal commented on src/plugins/discover/public/application/main/hooks/use_discover_state.ts:
```diff
@@ -220,12 +245,53 @@ export function useDiscoverState({
+            index: dataViewObj.id,
+            ...(columns.length && { columns }),
+          };
+          stateContainer.setAppState(nextState);
```
I think this should solve the back button problem @davismcphee mentioned.

So we're setting the columns after the SQL query is executed. And then we modify the app state, which pushes to history. With the following change history is replace, not pushed:

```suggestion
          stateContainer.replaceUrlAppState(nextState);
```

kertal replied:
sorry for this mess here, can't wait to refactor 😄

stratoula commented:
Things that need to be addressed on follow-up PRS have been added here https://github.com/elastic/kibana/issues/136950 cc @ghudgins

davismcphee reviewed (approved):
Looks like most of what I reported is already fixed and Discover code changes LGTM!

stratoula commented:
@elasticmachine merge upstream

stratoula commented on src/plugins/discover/public/application/main/components/sidebar/discover_field.tsx:
```diff
@@ -280,6 +281,7 @@ function DiscoverFieldComponent({
   showFieldStats,
 }: DiscoverFieldProps) {
   const [infoIsOpen, setOpen] = useState(false);
+  const isPlainRecord = !!onAddFilter;
```
Oh Dima made this change. I will change the variable name because it is confusing, thanx for noticing that

flash1293 reviewed (approved):
VisEditors changes look mostly good to me, noticed two small things but they don't block the PR:

Data view recognition is using the first `FROM` from the query, but in case of nested selects this doesn't work:
<img width="957" alt="Screenshot 2022-07-25 at 10 27 29" src="https://user-images.githubusercontent.com/1508364/180733704-5ff62f53-9ae2-4208-987f-5136cb3e8373.png">

It's possible to save a saved search, build a vis on it, then switch it to sql and save it again. The panel then fails like this:
<img width="667" alt="Screenshot 2022-07-25 at 10 32 00" src="https://user-images.githubusercontent.com/1508364/180733815-486a038f-879c-4a18-b228-dc7c2fb32b8f.png">
This could be fixed by "forgetting" the current saved search when switching to/from SQL mode.

Approving as I'm happy to merge as-is

stratoula commented:
Thanx @flash1293! I am moving them to the follow up issue!

stratoula commented:
Fixed the nested selects problem (wrong dataview recognition)

<img width="2484" alt="image" src="https://user-images.githubusercontent.com/17003240/180767587-cef9545e-9baa-4867-b334-7b5d6a04b134.png">

MichaelMarcialis reviewed (approved):
Thanks for making those additional requested changes, @stratoula. Amazing work! I've put together an updated list of comments for your review below.

I know feature freeze is looming and we're eager to get this feature in. At the end of the day, none of my comments are what I'd consider hard blockers, so I'm going to go ahead and approve now so I don't hold you up further (under the assumption that these either get addressed in this PR or spun off as a separate issue/PR).

# Replies

> About the resizer I am fine on copying the styles but it is a bit confusing for me as I am not very comfortable with the EUI code. I would need some help here.

Certainly. Do you want my assistance, or are you looking for one of the experts from the @elastic/eui-design team? If me, feel free to throw something on the calendar for us to hop on a Zoom.

# New Comments

- For the error popover, would it be possible to put an `alignItems="flexStart"` prop on the containing `EuiFlexGroup`? Doing so should prevent the first column (with icon and line number) from being vertical aligned to the center (noticeable when the message breaks a line).

![image](https://user-images.githubusercontent.com/3884767/180858543-e233ff0e-0cd3-4f25-84ff-6e8cdd2ebf9d.png)

- Would it be possible to support word breaking (via the `euiTextBreakWord` mixin, I believe) for the messages within the error popover? Currently, long strings of unbroken text can overflow out of the popover.

![image](https://user-images.githubusercontent.com/3884767/180859478-08e01982-1fb4-4d71-8fde-b40439f2b11b.png)

- For the unfocused editor in compact mode, the presence of the invalid red bottom border appears to be throwing off the editor's alignment (by increasing it's height by 1px) relative to other adjacent elements. Can this be remedied? If the custom styles currently in place aren't working, one alternative may be to use styles like EUI does for standard input (where both the focus and invalid colored borders are created via a background gradient).

![image](https://user-images.githubusercontent.com/3884767/180863972-f497ad17-e76b-421d-8065-4f08d52eea64.png)

- The editor footer in expanded mode still appears to be misaligned by 1px horizontally.

![image](https://user-images.githubusercontent.com/3884767/180864498-74d1bd40-5839-4550-aac8-05057ebeb87a.png)

- During a resize action, the gray border above the editor footer disappears and a blue border appears below/behind the footer. Can we correct this so that the gray border doesn't disappear and the blue border doesn't appear during a resize action?

![image](https://user-images.githubusercontent.com/3884767/180865320-019f2782-078e-4956-9f37-46babd1a2f0b.png)

- During a resize action, if an invalid border is present on the editor, it temporarily turns from a 2px border to a 1px border. After stopping the resize action, it correctly returns to 2px after a few seconds. Possible to fix?

![image](https://user-images.githubusercontent.com/3884767/180865097-64ef2247-774d-44db-ada8-45dfd3521ea4.png)

- It appears that field type filtering is being omitted when in SQL mode. Is this intentional?

- It appears that field type tokens and tooltips are missing from the field list in SQL mode. Is this intentional?

- Interacting with one or more "Remove field from table" buttons in the field list doesn't appear to alter the SQL query. For example, if the user's query selects all fields (via `SELECT *`), removing a field from the table keeps the wildcard (`*`) intact, which I found somewhat confusing. As a user, I'd expect the query to break out the wildcard into only the selected fields (i.e. all but the ones the user has chose to remove from the table). Is the current behavior intentional? Also worth noting with this currently implemented behavior is that refreshing doesn't restore all fields to the table, even though the wildcard is still present in the `SELECT` clause.

kertal commented:
> * It appears that field type filtering is being omitted when in SQL mode. Is this intentional?
> * It appears that field type tokens and tooltips are missing from the field list in SQL mode. Is this intentional?

yes, this is something to discuss. In SQL you can do the following:

`SELECT geoip.continent_name as "taxless_total_price", taxless_total_price as "some other name"`

In this case `taxless_total_price` would have the wrong type, and `taxless_total_price` would have an unknown type.
Secondly if you add aggregations, these would have unknown types. So the information about the types could be unreliable / or be primarily unknown ... if we are fine with this, it's simple to enable filtering and tokens, but it needs to be considered.  

> * Interacting with one or more "Remove field from table" buttons in the field list doesn't appear to alter the SQL query. For example, if the user's query selects all fields (via `SELECT *`), removing a field from the table keeps the wildcard (`*`) intact, which I found somewhat confusing. 

This is simply a limitation how Discover currently works , there is the data layer and the presentation layer, when you're adding/removing fields you're changing the presentation of your results and not the query. But it's not written in stone of course.

stratoula commented:
Thanx @MichaelMarcialis, I really appreciate it! I addressed everything except from the 2 issues that have to do with the resizer. I added them to the followup issue https://github.com/elastic/kibana/issues/136950. This has something to do with monaco height update so I need some time to investigate it and find alternatives.

ppisljar commented on x-pack/plugins/maps/public/legacy_visualizations/tile_map/tile_map_fn.ts:
```diff
@@ -45,14 +45,15 @@ export const createTileMapFn = (): TileMapExpressionFunctionDefinition => ({
     },
   },
   async fn(input, args) {
+    const query = input.query as Query;
```
how can we be sure that what's passed in here is not an AggregateQuery?

ppisljar commented on x-pack/plugins/reporting/public/panel_actions/get_csv_panel_action.tsx:
```diff
@@ -98,7 +97,20 @@ export class ReportingCsvPanelAction implements ActionDefinition<ActionContext>
+    }
+
+    const savedSearch = embeddable.getSavedSearch();
+    const query = savedSearch.searchSource.getField('query');
```
is it not possible that searchsource contains the query of type sql but it gets dropped (ignored) in the actual embeddable?

ppisljar commented on x-pack/plugins/reporting/public/panel_actions/get_csv_panel_action.tsx:
```diff
@@ -98,7 +97,20 @@ export class ReportingCsvPanelAction implements ActionDefinition<ActionContext>
+      return false;
+    }
+
+    const savedSearch = embeddable.getSavedSearch();
```
can we be sure here that embeddable contains a saved search ?

ppisljar commented on src/plugins/visualizations/public/visualize_app/utils/use/use_editor_updates.ts:
```diff
@@ -51,15 +51,15 @@ export const useEditorUpdates = (
             timeRange: timefilter.getTime(),
             filters: filterManager.getFilters(),
-            query: queryString.getQuery(),
+            query: queryString.getQuery() as Query,
```
can we be sure this is not an AggregateQuery ?
